explorer/interpreter/pattern_match.cpp

```diff
@@ -200,14 +201,37 @@ auto PatternMatch(Nonnull<const Value*> p, ExpressionResult v,
       // on the typechecker to ensure that `v.value()` is a type.
       return true;
     case Value::Kind::StaticArrayType: {
+      const auto& p_arr = cast<StaticArrayType>(*p);
       switch (v.value()->kind()) {
         case Value::Kind::TupleType:
         case Value::Kind::TupleValue: {
+          const auto& v_tup = cast<TupleValueBase>(*v.value());
+          if (v_tup.elements().empty()) {
+            return !TypeIsDeduceable(&p_arr.element_type());
+          }
+
+          const size_t deduced_size = GetSize(v_tup.elements()[0]);
+          for (const auto& tup_elem : v_tup.elements()) {
+            if (!PatternMatch(&p_arr.element_type(), make_expr_result(tup_elem),
+                              source_loc, bindings, generic_args, trace_stream,
+                              arena)) {
+              return false;
+            }
+
+            if (deduced_size != GetSize(tup_elem)) {
```

Comment:
I'm having trouble following this. If I understand correctly, this is trying to make sure that the tuple elements all have the same size, right? If so, that's confusing to me, because in this context it seems like we have no particular reason to expect the tuple elements to have a meaningful size, so this seems out of place.

Also, won't the type checker have already taken care of this? If not, don't we also need to check that `v_tup.elements().size()` matches the size of the array?

(Part of my confusion here is that we're using pattern matching to handle something that really should be type deduction, but that's a pre-existing problem.)

Reply:
Sorry for the delay with this.

> If I understand correctly, this is trying to make sure that the tuple elements all have the same size, right?

Indeed, the idea is to check that the tuple has a homogenous structure, thus matches the array structure. Could you please elaborate a bit on 

> we have no particular reason to expect the tuple elements to have a meaningful size

?
Do you mean that, e.g. i32 has no meaningful size? If so, `GetSize` returns 0 for such types. Probably it'd be more explicit to return `std::optional<size_t>` holding an actual value only for "sizeable" types?

> Also, won't the type checker have already taken care of this?

Probably I'm missing some contexts, where this function is also used, but AFAIU `TypeChecker::TypeCheckPattern` uses it for type deduction, so no checks are performed yet. 

>  If not, don't we also need to check that v_tup.elements().size() matches the size of the array?

Good point, I guess currently this check is performed during the implicit conversion of the tuple to the array and pattern matching is actually used for type deduction. However, for cases like
```
var x: [[i32;];3] = ((1,2), (3,4,5));
```
it seems like pattern matching should fail. I'll add the check.

Reply:
Ohhhh, I think I understand what's going on now: we're deducing the element type of an array from the initializer, so we need to deduce from each element separately, and verify that all deductions produce the same result. But since we're limiting this to the case where the element type is already known to be an array type, we only need to verify that all deductions deduce the same size.

Assuming that's right, I think the code will be clearer (and more general) if we don't rely on that limitation here: instead of checking that the deduced types all _have the same size_, I think we should check that the deduced types _are all equal_. Does that make sense?

(It might even be better to drop that limitation altogether, and add support for things like `var x: [auto; 3] = (1, 2, 3);`, but that doesn't necessarily need to be in this PR.)

Reply:
Yep, the general idea is correct. However we do not make any assumptions about array elements' type, we just know that the pattern we're trying to match with is an array.
So what's happening here is we check, that every tuple element matches array's element type.
However this is not enough, as if the array's elements are arrays themselves and their size should be deduced, then even if tuple elements have different sizes, they'll successfully match "outer" array's elements' type. Thus check for deduced size added. 

I agree, that code is not obvious and your suggestion absolutely makes sense.
To sum up:

1. Check that all tuple elements have equal types
2. Check that this type matches array's elements' type
3. If array's size explicitly set, check that it's equal to tuple's size

Does it look correct to you?

It also seems like it enables support for `auto` as an element type.

Reply:
That sounds basically right, but I was envisioning something subtly different for steps 1 and 2: rather than check the tuple element types are equal, we use each tuple element type to deduce the array element type, and then check that the results of those deductions are equal. That way the same logic should work for the case where the tuple elements have different types, but the array element type is explicitly specified, and the tuple element types are convertible to it.